src/__Private/ParserCache.hack

```diff
@@ -71,11 +72,47 @@ final class ParserCache {
     );
   }
 
-  private static function getCacheFileName(File $file): ?string {
+  private ?string $repoRoot = null;
+  const string NO_REPO_ROOT = "\0NO_REPO_ROOT";
+
+  private function getRepoRoot(string $file_path): ?string {
+    if ($this->repoRoot === self::NO_REPO_ROOT) {
+      return null;
+    }
+    if (
+      $this->repoRoot !== null && Str\starts_with($file_path, $this->repoRoot)
+    ) {
+      return $this->repoRoot;
+    }
+
+    $parts = Str\split($file_path, '/');
+    while (!C\is_empty($parts)) {
+      $dir = Str\join($parts, '/');
+      if (\file_exists($dir.'/.hhconfig')) {
+        $this->repoRoot = $dir;
+        return $dir;
+      }
+      $parts = Vec\take($parts, C\count($parts) - 1);
+    }
+    $this->repoRoot = self::NO_REPO_ROOT;
+    return null;
+  }
+
+  private function getCacheFileName(File $file): ?string {
     if ($file->isDirty()) {
       return null;
     }
+
     $source = $file->getPath();
+    $root = $this->getRepoRoot($source);
+    if ($root !== null) {
+      $dir = $root.'/.var/cache/hhvm/hhast';
```

Comment:
Maybe just `$root.'/.var/hhast-cache'`? Or do you expect each of the directories to contain other stuff in the future so the multi-level hierarchy is warranted?

Reply:
I was intending to use github orgs as a vendor ID to avoid conflicts